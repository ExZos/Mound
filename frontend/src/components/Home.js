import React from 'react';

import GeneralComponent from './GeneralComponent';
import Header from './Header';
import ConfirmModal from './ConfirmModal';
import { server, api } from '../server';
import '../styles/home.css';

class Home extends GeneralComponent {
  constructor(props) {
    super(props);

    this.state = {
      space: {
        name: ''
      }
    };
  }

  componentDidMount() {
    // TEMP
    console.log(this.getSessionItem('users'));
  }

  getSpaceByName = () => {
    server.get(api.getSpaceByName + this.state.space.name + '/')
      .then((res) => {
          this.setState({
            space: res.data
          });

          this.props.history.push({
            pathname: '/s/',
            state: {
              space: this.state.space
            }
          });
        })
      .catch((err) => {
        if(this.state.space.name) {
          this.toggleModal();
        }
      });
  }

  addSpace = () => {
    server.post(api.spaces, this.state.space)
      .then((res) => {
        this.setState({
          space: res.data
        });

        this.props.history.push({
          pathname: '/s/',
          state: {
            space: this.state.space
          }
        });
      });
  }

  render() {
    return (
      <div id="home">
        <Header />

        <br />

        <div>
          <form id="getSpaceByName" onSubmit={this.handleFormSubmit}>
            <input type="text" name="name" placeholder="Type a space name..." autoFocus
              value={this.state.space.name}
              onChange={(e) => this.handleInputChange(e, 'space')}
            />

            <button tabIndex="-1" onClick={this.getSpaceByName}>ENTER</button>
          </form>

          <ConfirmModal showModal={this.state.showModal}
            toggleModal={this.toggleModal} confirm={this.addSpace}
            mHeader={"Space '" + this.state.space.name + "' does not exist"}
            mBody="Do you want to request approval for this space?"
          />
        </div>
      </div>
    );
  }
}

export default Home;
