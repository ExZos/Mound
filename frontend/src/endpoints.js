export const endpoints = {
  messages: '/api/messages/',
  spaces: '/api/spaces/',
  users: '/api/users/',
  poll: '/api/polls/',
  votes: '/api/votes/',
  getSpaceByName: '/space/getByName/',
  getUserCountInSpaceForUser: '/space/getUserCountForUser/',
  getUserByName: '/user/getByName/',
  getUsersInSpace: '/user/getInSpace/',
  getUserInSpaceByName: '/user/getInSpaceByName/',
  getUsersInSpaceExceptName: '/user/getInSpaceExceptName/',
  createUserNApproveSpace: 'user/createNApproveSpace/',
  getMessagesInSpace: '/message/getInSpace/',
  getPendingPollsInSpace: '/poll/getPendingInSpace/',
  getPendingUnvotedPollsForUser: '/poll/getPendingUnvotedForUser/',
  getPendingJoinPollInSpaceByName: 'poll/getPendingJoinInSpaceByName/',
  getJoinPollResults: '/poll/getJoinResults/',
  createNameRelatedPoll: '/poll/createNameRelated/',
  getVotesForPoll: 'vote/getForPoll/',
  getPositiveVotesForPoll: 'vote/getPositivesForPoll/',
  createVoteNUpdatePoll: 'vote/createNUpdatePoll/',
}
