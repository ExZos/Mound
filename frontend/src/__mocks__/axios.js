import { endpoints as api } from '../endpoints';

const axios = {
  create: () => axios,
  get: (url) => {
    switch(url) {
      case api.getSpaceByName + "Headon's Floor/":
        return Promise.resolve({
          data: {
            id: 1,
            name: "Headon's Floor",
            status: true
          }
        });
        break;

      default:
        return Promise.reject();
    }
  }
};

export default axios;
