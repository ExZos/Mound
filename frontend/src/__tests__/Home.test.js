import React from 'react';
import { shallow, mount } from 'enzyme';
import axios from 'axios';
import { createBrowserHistory } from 'history';

import Home from '../components/Home';
import { endpoints as api } from '../endpoints';

const history = createBrowserHistory();

describe('Home', () => {
  it('should render correctly with no props', () => {
    const component = shallow(<Home />);

    expect(component).toMatchSnapshot();
  });

  it('should create a state entry equal to the input value', () => {
    const component = shallow(<Home />);
    const input = component.find('input');

    input.props().onChange({
      target: {
        name: 'name',
        value: 'space1'
      }
    });

    expect(component.state('space')).toBeDefined();
    expect(component.state('space')).toHaveProperty('name', 'space1');
  });

  it('should create a state entry on valid submit', async () => {
    const component = shallow(<Home history={history} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    component.setState({
      space: {
        name: 'space1'
      }
    });
    await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getSpaceByName + 'space1/');
    expect(component.state('space')).toBeDefined();
    expect(component.state('space')).toHaveProperty('id');
  });

  it('should not create a state entry on invalid submit and should trigger modal', async () => {
    const component = shallow(<Home history={history} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    component.setState({
      space: {
        name: 'Missing Space'
      }
    });
    await await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getSpaceByName + 'Missing Space/');
    expect(component.state('space')).toBeDefined();
    expect(component.state('space')).not.toHaveProperty('id');
    expect(component.state('showModal')).toBeDefined();
    expect(component.state('showModal')).toBeTruthy();
  });

  it('should not create a state entry nor trigger modal on empty submit', async () => {
    const component = shallow(<Home history={history} />);
    const button = component.find('button');
    const spy = jest.spyOn(axios, 'get');

    component.setState({
      space: {
        name: ''
      }
    });
    await await button.props().onClick();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.getSpaceByName + '/');
    expect(component.state('space')).toBeDefined();
    expect(component.state('space')).not.toHaveProperty('id');
    expect(component.state('showModal')).toBeDefined();
    expect(component.state('showModal')).toBeFalsy();
  });

  it('should call post api on modal confirm', () => {
    const component = shallow(<Home history={history} />);
    const confirmDialog = component.find('ConfirmDialog');
    const spy = jest.spyOn(axios, 'post');

    component.setState({
      space: {
        name: 'space1'
      }
    });
    confirmDialog.props().confirm();

    expect(spy).toHaveBeenCalled();
    expect(spy).toHaveBeenLastCalledWith(api.spaces, {
      name: 'space1'
    });
  });
});
