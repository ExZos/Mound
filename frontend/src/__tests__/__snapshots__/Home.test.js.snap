// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly with no props 1`] = `
<div
  id="home"
>
  <Sidebar />
  <br />
  <form
    id="getSpaceByName"
    onSubmit={[Function]}
  >
    <WithStyles(ForwardRef(ButtonGroup))
      size="small"
      variant="outlined"
    >
      <WithStyles(ForwardRef(TextField))
        autoFocus={true}
        label="Type a space name..."
        name="name"
        onChange={[Function]}
        value=""
      />
      <WithStyles(ForwardRef(Button))
        color="primary"
        onClick={[Function]}
        tabIndex="-1"
        type="submit"
      >
        ENTER
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(ButtonGroup))>
  </form>
  <ConfirmDialog
    confirm={[Function]}
    mBody="Do you want to request approval for this space?"
    mHeader="Space '' does not exist"
    showModal={false}
    toggleModal={[Function]}
  />
</div>
`;
